File: batch-tech-feed/produce/techfeed/publish.py
"""焼き上がった mp4 を pechka の配信経路に載せる。

    mp4 ──▶ HLS 変換 ──▶ MinIO ──▶ PostgreSQL ──▶ (Benthos CDC) ──▶ MongoDB / Elasticsearch
                                                                        │
                                                                        ▼
                                                                    pechka で再生

`contents.status` は最後に `ready` へ上げる。アップロードの途中で落ちても、
中途半端なコンテンツが一覧に出ないようにするためである。
"""

from __future__ import annotations

import tempfile
from pathlib import Path

from . import catalog, hls, storage


def run(
    manifest: dict,
    mp4: str,
    source_key: str,
    description: str = "",
    keep_work: str | None = None,
) -> dict:
    if not Path(mp4).is_file():
        raise SystemExit(f"{mp4} not found; run the 'build' step first")

    title = manifest.get("title") or "技術ダイジェスト"
    duration_seconds = round(manifest.get("total_ms", 0) / 1000)

    store = storage.Storage.from_env()
    conn = catalog.connect()
    try:
        content_id, short_id = catalog.upsert_content(
            conn, source_key, title, description, duration_seconds
        )
        print(f"content: id={content_id} short_id={short_id} source_key={source_key}")

        with tempfile.TemporaryDirectory(prefix="techfeed-hls-") as tmp:
            work = keep_work or tmp
            Path(work).mkdir(parents=True, exist_ok=True)

            print("transcoding to HLS...")
            variants = hls.transcode(mp4, work)

            prefix = storage.hls_prefix(short_id)
            print(f"uploading HLS to s3://{store.bucket}/{prefix}/ ...")
            uploaded = store.put_dir(work, prefix)
            store.put_text(hls.master_playlist(variants), f"{prefix}/master.m3u8")
            print(f"  {uploaded} file(s) + master.m3u8")

            thumb = Path(work) / "thumb_01.jpg"
            hls.thumbnail(mp4, str(thumb))
            thumb_key = storage.thumbnail_key(short_id)
            store.put_file(str(thumb), thumb_key)
            catalog.set_thumbnail(conn, content_id, thumb_key)
            print(f"  thumbnail -> {thumb_key}")

        # master も再生対象として登録する。プレーヤーは既定でこれを掴む。
        catalog.register_variant(conn, content_id, short_id, "master", None, None, None)
        for v in variants:
            catalog.register_variant(
                conn, content_id, short_id, v.name, v.bandwidth, v.resolution, v.codecs
            )
        print(f"registered {len(variants) + 1} variant(s)")

        catalog.mark_ready(conn, content_id)
        print(f"content is ready: /contents/{short_id}")
    finally:
        conn.close()

    return {"content_id": content_id, "short_id": short_id}

File: batch-tech-feed/produce/cmd/publish.go
package cmd

import (
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"math"
	"os"
	"path/filepath"

	"github.com/k-wa-wa/pechka/batch-tech-feed/produce/shared"
)

func RunPublish(ctx context.Context, osArgs []string) error {
	fs := flag.NewFlagSet("publish", flag.ContinueOnError)
	outDir := fs.String("out", "", "working directory")
	output := fs.String("output", "", "mp4 to publish (default: <out>/digest.mp4)")
	description := fs.String("description", "", "content description")

	if err := fs.Parse(osArgs); err != nil {
		return err
	}

	if *outDir == "" {
		return fmt.Errorf("--out parameter is required")
	}

	manifestPath := filepath.Join(*outDir, "manifest.json")
	raw, err := os.ReadFile(manifestPath)
	if err != nil {
		return fmt.Errorf("%s not found; run the 'synthesize' step first", manifestPath)
	}

	var manifest shared.Manifest
	if err := json.Unmarshal(raw, &manifest); err != nil {
		return fmt.Errorf("failed to parse manifest: %w", err)
	}

	mp4Path := *output
	if mp4Path == "" {
		mp4Path = filepath.Join(*outDir, "digest.mp4")
	}

	if _, err := os.Stat(mp4Path); err != nil {
		return fmt.Errorf("%s not found; run the 'build' or 'render' step first", mp4Path)
	}

	dateStr := manifest.DigestDate
	if dateStr == "" {
		dateStr = "unknown"
	}
	sk := fmt.Sprintf("tech-feed:%s", dateStr)

	title := manifest.Title
	if title == "" {
		title = "技術ダイジェスト"
	}
	durationSec := int(math.Round(float64(manifest.TotalMs) / 1000.0))

	store, err := shared.NewStorageFromEnv(ctx)
	if err != nil {
		return fmt.Errorf("failed to init storage: %w", err)
	}

	conn, err := shared.ConnectCatalogDB(ctx)
	if err != nil {
		return fmt.Errorf("failed to connect to DB: %w", err)
	}
	defer conn.Close(ctx)

	contentID, shortID, err := shared.UpsertContent(ctx, conn, sk, title, *description, durationSec, nil)
	if err != nil {
		return fmt.Errorf("failed to upsert content: %w", err)
	}
	fmt.Printf("content: id=%s short_id=%s source_key=%s\n", contentID, shortID, sk)

	tmpWorkDir, err := os.MkdirTemp("", "techfeed-hls-")
	if err != nil {
		return fmt.Errorf("failed to create temp dir: %w", err)
	}
	defer os.RemoveAll(tmpWorkDir)

	fmt.Println("transcoding to HLS...")
	variants, err := shared.TranscodeHLS(mp4Path, tmpWorkDir)
	if err != nil {
		return fmt.Errorf("failed to transcode HLS: %w", err)
	}

	prefix := shared.HLSPrefix(shortID)
	fmt.Printf("uploading HLS to s3://%s/%s/ ...\n", store.Bucket, prefix)
	count, err := store.PutDir(ctx, tmpWorkDir, prefix)
	if err != nil {
		return fmt.Errorf("failed to upload HLS dir: %w", err)
	}

	if err := store.PutText(ctx, shared.MasterPlaylist(variants), fmt.Sprintf("%s/master.m3u8", prefix)); err != nil {
		return fmt.Errorf("failed to put master.m3u8: %w", err)
	}
	fmt.Printf("  %d file(s) + master.m3u8\n", count)

	thumbPath := filepath.Join(tmpWorkDir, "thumb_01.jpg")
	if _, err := shared.GenerateThumbnail(mp4Path, thumbPath, 2.0); err != nil {
		return fmt.Errorf("failed to generate thumbnail: %w", err)
	}

	thumbKey := shared.ThumbnailKey(shortID)
	if err := store.PutFile(ctx, thumbPath, thumbKey); err != nil {
		return fmt.Errorf("failed to upload thumbnail: %w", err)
	}

	if err := shared.SetThumbnail(ctx, conn, contentID, thumbKey); err != nil {
		return fmt.Errorf("failed to set thumbnail in DB: %w", err)
	}
	fmt.Printf("  thumbnail -> %s\n", thumbKey)

	// register master variant
	if err := shared.RegisterVariant(ctx, conn, contentID, shortID, "master", nil, nil, nil); err != nil {
		return fmt.Errorf("failed to register master variant: %w", err)
	}

	for _, v := range variants {
		var bw *int
		var res *string
		var c *string
		if v.Bandwidth > 0 {
			val := v.Bandwidth
			bw = &val
		}
		if v.Resolution != "" {
			val := v.Resolution
			res = &val
		}
		if v.Codecs != "" {
			val := v.Codecs
			c = &val
		}
		if err := shared.RegisterVariant(ctx, conn, contentID, shortID, v.Name, bw, res, c); err != nil {
			return fmt.Errorf("failed to register variant %s: %w", v.Name, err)
		}
	}
	fmt.Printf("registered %d variant(s)\n", len(variants)+1)

	if err := shared.MarkReady(ctx, conn, contentID); err != nil {
		return fmt.Errorf("failed to mark content ready: %w", err)
	}
	fmt.Printf("content is ready: /contents/%s\n", shortID)

	return nil
}
